File: app/src/main/java/com/blog/demo/md/SnackBarActivity.java
package com.blog.demo.md;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.view.View;
import android.widget.Toast;

import com.blog.demo.LogTool;
import com.blog.demo.R;

public class SnackBarActivity extends Activity implements View.OnClickListener {
    private static final String LOG_TAG = "SnackBarActivity";
    private Snackbar mSnackbar;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_material_design_snack_bar);

        findViewById(R.id.btn_cancel).setOnClickListener(this);
        findViewById(R.id.btn_in_coordinator).setOnClickListener(this);
        findViewById(R.id.btn_in_relative).setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.btn_cancel) {
            if (mSnackbar != null) {
                mSnackbar.dismiss();
            }
        } else {
            mSnackbar = Snackbar.make(v, "Snackbar", Snackbar.LENGTH_LONG);
            mSnackbar.show();

            if (v.getId() == R.id.btn_in_coordinator) {
                mSnackbar.setAction("Confirm", new View.OnClickListener(){
                    @Override
                    public void onClick(View v) {
                        Toast.makeText(SnackBarActivity.this, "onClick",
                                Toast.LENGTH_LONG).show();
                    }
                });

                mSnackbar.addCallback(new Snackbar.Callback(){
                    public void onShown(Snackbar sb) {
                        LogTool.logi(LOG_TAG, "onShown");
                    }

                    public void onDismissed(Snackbar transientBottomBar, int event) {
                        if (event == Snackbar.Callback.DISMISS_EVENT_SWIPE) {
                            LogTool.logi(LOG_TAG, "swipe");
                        } else if (event == Snackbar.Callback.DISMISS_EVENT_ACTION) {
                            LogTool.logi(LOG_TAG, "action");
                        } else if (event == Snackbar.Callback.DISMISS_EVENT_TIMEOUT) {
                            LogTool.logi(LOG_TAG, "timeout");
                        } else if (event == Snackbar.Callback.DISMISS_EVENT_MANUAL) {
                            LogTool.logi(LOG_TAG, "manual");
                        } else if (event == Snackbar.Callback.DISMISS_EVENT_CONSECUTIVE) {
                            LogTool.logi(LOG_TAG, "consecutive");
                        }

                    }
                });
            }
        }
    }

}

File: app/src/main/java/com/blog/demo/md/SnackBarActivity.kt
package com.blog.demo.md

import android.app.Activity
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.Toast
import com.blog.demo.LogTool.logi
import com.blog.demo.R
import com.google.android.material.snackbar.Snackbar

class SnackBarActivity : Activity(), View.OnClickListener {

    private val LOG_TAG = "SnackBarActivity"
    private var mSnackbar: Snackbar? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_material_design_snack_bar)

        findViewById<Button>(R.id.btn_cancel).setOnClickListener(this)
        findViewById<Button>(R.id.btn_in_coordinator).setOnClickListener(this)
        findViewById<Button>(R.id.btn_in_relative).setOnClickListener(this)
    }

    override fun onClick(v: View) {
        if (v.id == R.id.btn_cancel) {
            mSnackbar?.dismiss()
        } else {
            mSnackbar = Snackbar.make(v, "Snackbar", Snackbar.LENGTH_LONG)
            mSnackbar?.show()

            if (v.id == R.id.btn_in_coordinator) {
                mSnackbar?.setAction("Confirm") {
                    Toast.makeText(this@SnackBarActivity, "onClick", Toast.LENGTH_LONG).show()
                }
                mSnackbar?.addCallback(object : Snackbar.Callback() {
                    override fun onShown(sb: Snackbar) {
                        logi(LOG_TAG, "onShown")
                    }

                    override fun onDismissed(transientBottomBar: Snackbar, event: Int) {
                        if (event == DISMISS_EVENT_SWIPE) {
                            logi(LOG_TAG, "swipe")
                        } else if (event == DISMISS_EVENT_ACTION) {
                            logi(LOG_TAG, "action")
                        } else if (event == DISMISS_EVENT_TIMEOUT) {
                            logi(LOG_TAG, "timeout")
                        } else if (event == DISMISS_EVENT_MANUAL) {
                            logi(LOG_TAG, "manual")
                        } else if (event == DISMISS_EVENT_CONSECUTIVE) {
                            logi(LOG_TAG, "consecutive")
                        }
                    }
                })
            }
        }
    }

}